src/nimble.nim

```diff
@@ -569,17 +579,35 @@ proc search(options: Options) =
 
   if not found:
     display("Error", "No package found.", Error, HighPriority)
+  elif options.jsonOutput:
+    echo jsonOutput
 
-proc list(options: Options) =
-  if needsRefresh(options):
-    raise newException(NimbleError, "Please run nimble refresh.")
-  let pkgList = getPackageList(options)
+proc echoList(pkgList: seq[Package], queryVersions: bool) =
   for pkg in pkgList:
     echoPackage(pkg)
-    if pkg.alias.len == 0 and options.queryVersions:
+    if pkg.alias.len == 0 and queryVersions:
       echoPackageVersions(pkg)
     echo(" ")
 
+proc jsonList(pkgList: seq[Package], queryVersions: bool) =
+  echo "["
+  for i, pkg in pkgList:
+    let info = pkg.json
+    if queryVersions:
+      info["versions"] = packageVersionsJson(pkg)
+    echo info
+    if i != pkgList.high: echo ","
+  echo "]"
```

Comment:
Please create a JArray and echo it.

Reply:
The tradeoff is clean code or streaming output, but streaming output is not a requirement.

Reply:
I'd prefer cleaner code.